### PR TITLE
[CALCITE-3366] RelDecorrelator supports Union (Xu Zhaohui & Jin Xing)

### PR DESCRIPTION
**Pull request by xy2953396112**
https://issues.apache.org/jira/browse/CALCITE-3366

### CODE REVIEW

**Comment by xy2953396112** [comment]
@hsyuan 
Please help review this pr when you have time, thanks.

File: core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml [hunk 1]
```diff
@@ -11699,6 +11699,249 @@ LogicalProject(JOB=[$1])
...
+                  LogicalProject(ENAME=[$1])
+                    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+            LogicalJoin(condition=[true], joinType=[inner])
+              LogicalProject(EXPR$0=[1], NAME0=[$2])
+                LogicalFilter(condition=[=(CAST($1):VARCHAR(20) NOT NULL, $2)])
+                  LogicalProject(DEPTNO=[$0], NAME=[$1], NAME0=[CAST($1):VARCHAR(20) NOT NULL])
+                    LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+              LogicalAggregate(group=[{0}])
+                LogicalProject(DEPTNO=[$7])
+                  LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```
Comment (vlsi) [inline core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:11883]: This looks sad :(

Have you considered decorellating as follows?
In other words, make distinct "core variables" produce outer joins with overall where condition that selects at least one of them.

```sql
SELECT *
  FROM emp e
  LEFT JOIN (select deptno from dept group by deptno) d1 on (d.deptno = e.deptno)
  LEFT JOIN (select name from dept group by name) d2 on (d.name = e.ename)
WHERE d1.deptno is not null
   or d2.name is not null
```

Comment (zabetak) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:11883]: Hey @vlsi, it seems that you have a performance concern regarding this PR. I didn't start reviewing yet but I was wondering if it makes sense to merge it even if the decorrelated plan is not the most efficient one. Some systems do not have a way to execute correlated queries so even if the decorrelated plan is not the best it is still some progress.

Comment (jamesstarr) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:11883]: Maybe a flag could be added that was off by default.

Comment (jamesstarr) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:11883]: > This looks sad :(
> 
> Have you considered decorellating as follows?
> In other words, make distinct "corellation variables" produce outer joins with overall where condition that selects at least one of them.
> 
> ```sql
> SELECT *
>   FROM emp e
>   LEFT JOIN (select deptno from dept group by deptno) d1 on (d.deptno = e.deptno)
>   LEFT JOIN (select name from dept group by name) d2 on (d.name = e.ename)
> WHERE d1.deptno is not null
>    or d2.name is not null
> ```

This optimization does not seem generalizable to a broader set queries.  For instance, if an aggregate function was between the union and the correlated rel.  The decorrelators first job is to provide a simple correct, logically correct expression of the query, and optimization can come later or better yet be generalized rules that do not **only** work on correlated queries.

File: core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java [hunk 2]
```diff
@@ -1272,6 +1274,138 @@ private static boolean isWidening(RelDataType type, RelDataType type1) {
...
+
+    final List<Frame> inputFrames = new ArrayList<>();
+    final Set<CorDef> corDefSet = new HashSet<>();
+
+    // Decorrelate to collect input frames and collect all
+    // the corDefOutputs from the frames. Note that the
+    // corDefOutputs might be different among the frames.
+    for (RelNode oldInput: oldInputs) {
+      final Frame frame = getInvoke(oldInput, isCorVarDefined, rel);
+      if (frame == null) {
```
Comment (jaystarshot) [inline core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:1294]: What if one part  of the union is changed due to de-correlation and other part is not changed ? why do we return null here even if one part of the union has changed? 
Eg - one side has Value Rel (hence returns null) , other has Correlate-...-TableScan

**Pull request closed without merging**